File: backend-java/patient-service/src/main/java/com/cdm/patient/controller/HealthMetricController.java
package com.cdm.patient.controller;

import com.cdm.common.security.IdentityPayload;
import com.cdm.patient.entity.HealthMetricEntity;
import com.cdm.patient.service.HealthMetricService;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.web.bind.annotation.*;
import java.util.Base64;
import java.util.List;

@RestController
@RequestMapping("/api/v1/health-metrics")
public class HealthMetricController {

    private final HealthMetricService service;

    public HealthMetricController(HealthMetricService service) {
        this.service = service;
    }

    // A basic stub for resolving identity from header
    private IdentityPayload getIdentity(String base64Identity) {
        try {
            String json = new String(Base64.getDecoder().decode(base64Identity));
            return new ObjectMapper().readValue(json, IdentityPayload.class);
        } catch (Exception e) {
            throw new RuntimeException("Invalid identity header", e);
        }
    }

    @GetMapping("/{patientId}")
    public List<HealthMetricEntity> getMetrics(
            @RequestHeader("X-Identity-Base64") String identityHeader,
            @PathVariable Long patientId
    ) {
        IdentityPayload identity = getIdentity(identityHeader);
        return service.findAllForPatient(identity, patientId);
    }

    @PostMapping("/{patientId}")
    public HealthMetricEntity createMetric(
            @RequestHeader("X-Identity-Base64") String identityHeader,
            @PathVariable Long patientId,
            @RequestParam String metricType,
            @RequestParam String metricValue
    ) {
        IdentityPayload identity = getIdentity(identityHeader);
        return service.create(identity, patientId, metricType, metricValue);
    }
}

File: backend-nestjs/apps/patient/src/health-metric/health-metric.controller.ts
import { Controller } from '@nestjs/common';
import { MessagePattern, Payload } from '@nestjs/microservices';
import { HealthMetricService } from './health-metric.service.js';
import type { PatientIdPayload, CreateHealthMetricPayload } from '@cdm/shared';

@Controller()
export class HealthMetricController {
  constructor(private readonly service: HealthMetricService) {}

  @MessagePattern({ cmd: 'health_metric_find_all' })
  async findAll(@Payload() data: PatientIdPayload) {
    return this.service.findAllForPatient(data.identity, data.patientId);
  }

  @MessagePattern({ cmd: 'health_metric_create' })
  async create(@Payload() data: CreateHealthMetricPayload) {
    return this.service.create(data.identity, data.patientId, { metricType: data.metricType, metricValue: data.metricValue });
  }
}
